src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/CustomNumberInput.razor

```diff
@@ -0,0 +1,20 @@
+﻿@using System.Diagnostics.CodeAnalysis;
+@inherits Microsoft.AspNetCore.Components.Forms.InputBase<int>
+
+<input name="@NameAttributeValue" @attributes="AdditionalAttributes" type="text" class="@CssClass" @bind="CurrentValueAsString" />
+
+@code {
+    protected override bool TryParseValueFromString(string value, [MaybeNullWhen(false)] out int result, [NotNullWhen(false)] out string validationErrorMessage)
+    {
+        if(int.TryParse(value, out result))
+        {
+            validationErrorMessage = null;
+            return true;
+        }else
+        {
+            // This is not used, since the error comes from the binding process.
```

Comment:
Trying to make sense of this. I would have thought something like:

 * If the model binding system registers an error for this field, that error will show up in the UI
 * Independently of that, if `CustomNumberInput` registers an error for this field, that error will show up in the UI

As such, either or both messages could appear. Is that the case?

In your E2E test case, does this method actually get executed?

 * If so it would be good to show that any messages returned from here can be seen in the UI
 * Or if not, it would be good to change this method to `throw new NotImplementedException` with a comment saying that the E2E scenario proves that we're only doing form-post binding and not interactive binding.